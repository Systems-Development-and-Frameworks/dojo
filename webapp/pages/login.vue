<template>
  <main>
    <LoginForm/>
  </main>
</template>

<script>
export default {
  layout: "default"
}
</script>

