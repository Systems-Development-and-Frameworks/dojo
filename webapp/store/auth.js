import jwt_decode from "jwt-decode"

export const state = () => ({
  token: null,
  userId: null,
})

export const getters = {
  isAuthenticated (state) {
    return !!state.token
  },
  userId (state) {
    return state.userId
  },
  token (state) {
    return state.token
  }
}

export const mutations = {
  _setToken (state, token) {
    state.token = token
  },
  _setUserId (state, userId) {
    state.userId = userId
  },
  unsetToken (state) {
    state.userId = null
    state.token = null
  }
}

export const actions = {
  async setToken (context, token) {
    const { userId } = jwt_decode(token)
    context.commit('_setToken', token)
    context.commit('_setUserId', userId)
  }
}
