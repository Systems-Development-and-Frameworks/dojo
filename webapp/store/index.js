import cookie from 'cookie'

export const actions = {
  async nuxtServerInit (store, context) {
    const { req } = context.ssrContext
    if (!req || typeof req.headers.cookie != 'string') return // static site generation
    const parsedCookies = cookie.parse(req.headers.cookie)
    const token = parsedCookies['apollo-token']
    if (!token) return
    req.headers.authorization = token
    return await store.dispatch('auth/setToken', token)
  },
}
