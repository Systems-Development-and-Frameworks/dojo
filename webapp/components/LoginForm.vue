<template>
  <div>
    <template v-if="isAuthenticated">
      <span>You're logged in!</span>
      <nuxt-link to="/">Return to home page</nuxt-link>
      or
      <a href="#" @click.prevent="$router.go(-1)">return to the previously browsed page</a>
    </template>
    <form v-else @submit.prevent="onSubmit" id="form-login">
      <table>
        <tbody>
        <tr>
          <td>
            <label for="input-email"> E-Mail </label>
          </td>
          <td>
            <input id="input-email" type="text" v-model.trim="email" placeholder="Enter your e-mail address.."
                   required/>
          </td>
        </tr>
        <tr>
          <td>
            <label for="input-password"> Password </label>
          </td>
          <td>
            <input id="input-password" type="password" v-model="password" placeholder="Enter your password.."
                   required/>
          </td>
        </tr>
        <tr>
          <td colspan="2">
            <input type="submit" value="Login" :disabled="!(email.length && password.length)"/>
          </td>
        </tr>
        <tr v-if="submitting">
          <td colspan="2">
            Logging in...
          </td>
        </tr>
        <tr v-if="error">
          <td colspan="2">
            <span id="login-error-message">{{ error }}</span>
          </td>
        </tr>
        </tbody>
      </table>
    </form>
  </div>
</template>

<script>
import login from '../apollo/mutations/login'
import { mapActions, mapGetters } from 'vuex'
import { ApolloError } from '@apollo/client'

export default {
  data () {
    return {
      submitting: false,
      error: null,
      email: '',
      password: '',
    }
  },
  computed: {
    ...mapGetters({
      isAuthenticated: 'auth/isAuthenticated'
    })
  },
  methods: {
    ...mapActions({
      setToken: 'auth/setToken'
    }),
    async onSubmit () {
      this.error = null
      this.submitting = true
      try {
        const res = await this.$apollo.mutate({
          mutation: login,
          variables: {
            email: this.email,
            password: this.password
          }
        }).then(({ data }) => data && data.login)
        await this.setToken(res)
        await this.$apolloHelpers.onLogin(res)
        if (this.$route.params.returnPath) {
          await this.$router.push({
            path: this.$route.params.returnPath
          })
        }
      } catch (e) {
        if (e instanceof ApolloError) {
          this.error = e.graphQLErrors.join(",\n")
        } else {
          this.error = e
        }
      } finally {
        this.submitting = false
      }
    }
  }
}
</script>

<style>
form#form-login > table {
  margin: 0 auto;
}
</style>

