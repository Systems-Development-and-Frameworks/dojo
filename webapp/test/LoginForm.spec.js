import '@testing-library/jest-dom'
import 'regenerator-runtime'

import { createLocalVue, shallowMount } from '@vue/test-utils'

import VueApollo from 'vue-apollo'
import { createMockClient } from 'mock-apollo-client'
import login from '@/apollo/mutations/login'

import Vuex from 'vuex'

import LoginForm from '@/components/LoginForm'
import { actions, getters, mutations, state } from '@/store/auth'
import { InMemoryCache } from '@apollo/client'

import jwt_decode from 'jwt-decode'

jest.mock('jwt-decode')

describe('LoginForm', () => {
  const localVue = createLocalVue()
  localVue.use(VueApollo)
  localVue.use(Vuex)

  let wrapper, mockClient, apolloProvider, store

  const createLoginForm = async (handlers, appOptions) => {
    mockClient = createMockClient({ cache: new InMemoryCache() })
    apolloProvider = new VueApollo({
      defaultClient: mockClient
    })

    const requestHandlers = {
      loginHandler: null,
      ...handlers
    }
    mockClient.setRequestHandler(
      login,
      requestHandlers.loginHandler
    )

    store = new Vuex.Store({
      modules: {
        auth: {
          namespaced: true,
          state,
          getters,
          actions,
          mutations
        }
      }
    })

    wrapper = shallowMount(LoginForm, {
      localVue,
      apolloProvider,
      store,
      stubs: {
        NuxtLink: true,
      },
      ...appOptions
    })
    await wrapper.vm.$nextTick()
  }

  afterEach(() => {
    wrapper.destroy()
    mockClient = null
    apolloProvider = null
    store = null
  })

  describe('when no user is logged in', () => {
    it('should show the login form', async () => {
      await createLoginForm()
      await wrapper.vm.$nextTick()

      expect(wrapper.find('#form-login').element).not.toBeEmptyDOMElement()
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('should only enable the submit button if both email and password are provided', async () => {
      await createLoginForm()
      await wrapper.vm.$nextTick()

      const loginButton = wrapper.find('input[value="Login"]').element

      expect(loginButton).toBeDisabled()

      wrapper.find('#input-email').setValue('someEmail@addr.es')
      await wrapper.vm.$nextTick()

      expect(loginButton).toBeDisabled()

      wrapper.find('#input-password').setValue('someSecret')
      await wrapper.vm.$nextTick()

      expect(loginButton).toBeEnabled()

      wrapper.find('#input-password').setValue('')
      await wrapper.vm.$nextTick()

      expect(loginButton).toBeDisabled()
    })
  })

  describe('on login', () => {
    jwt_decode.mockReturnValue({ userId: '1234' })

    const loginHandler = jest.fn().mockResolvedValue({ data: { login: 'someToken' } })

    let $router, $route, $apolloHelpers
    beforeEach(() => {
      $router = {
        push () {
        }
      }
      $route = { params: { returnPath: undefined } }
      $apolloHelpers = {
        onLogin: jest.fn()
      }
    })

    const login = async () => {
      wrapper.find('#input-email').setValue('someEmail@addr.es')
      wrapper.find('#input-password').setValue('someSecret')
      await wrapper.vm.$nextTick()

      await wrapper.find('#form-login').trigger('submit')
      await wrapper.vm.$nextTick()
    }

    it('should call onSubmit', async () => {
      await createLoginForm({ loginHandler }, {
        mocks: {
          $apolloHelpers,
          $router,
          $route
        }
      })
      await wrapper.vm.$nextTick()

      const spy = jest.spyOn(wrapper.vm, 'onSubmit')

      await login()

      expect(spy).toBeCalledTimes(1)
    })

    it('should send a login mutation with the email and password', async () => {
      await createLoginForm({ loginHandler }, {
        mocks: {
          $apolloHelpers,
          $router,
          $route
        }
      })
      await wrapper.vm.$nextTick()

      await login()

      expect(loginHandler).toHaveBeenNthCalledWith(1, {
        email: 'someEmail@addr.es',
        password: 'someSecret'
      })
    })

    describe('that succeeds', () => {
      it('should set the token and user id in the state', async () => {
        await createLoginForm({ loginHandler }, {
          mocks: {
            $apolloHelpers,
            $router,
            $route
          }
        })
        await wrapper.vm.$nextTick()

        await wrapper.vm.onSubmit()

        expect(store.getters['auth/isAuthenticated']).toBeTruthy()
        expect(store.getters['auth/userId']).toBe('1234')
        expect(store.getters['auth/token']).toBe('someToken')
      })

      it('should call $apolloHelpers.onLogin', async () => {
        await createLoginForm({ loginHandler }, {
          mocks: {
            $apolloHelpers,
            $router,
            $route
          }
        })
        await wrapper.vm.$nextTick()

        await wrapper.vm.onSubmit()

        expect($apolloHelpers.onLogin).toHaveBeenNthCalledWith(1, 'someToken')
      })

      it('should not push a route without given a returnPath', async () => {
        const routerPush = jest.spyOn($router, 'push')
        await createLoginForm({ loginHandler }, {
          mocks: {
            $apolloHelpers,
            $router,
            $route
          }
        })
        await wrapper.vm.$nextTick()

        await wrapper.vm.onSubmit()

        expect(routerPush).toHaveBeenCalledTimes(0)
      })

      it('should push a route given a returnPath', async () => {
        $route.params.returnPath = 'abc'
        const routerPush = jest.spyOn($router, 'push')
        await createLoginForm({ loginHandler }, {
          mocks: {
            $apolloHelpers,
            $router,
            $route
          }
        })
        await wrapper.vm.$nextTick()

        await wrapper.vm.onSubmit()

        expect(routerPush).toHaveBeenNthCalledWith(1, {
          path: 'abc'
        })
      })
    })

    describe('that fails', () => {
      it('should display an error', async () => {
        await createLoginForm({
          loginHandler: jest.fn().mockResolvedValue({
            data: null,
            errors: ['Some error']
          })
        }, {
          mocks: {
            $apolloHelpers,
            $router,
            $route
          }
        })
        await wrapper.vm.$nextTick()

        await wrapper.vm.onSubmit()

        await wrapper.vm.$nextTick()

        expect(wrapper.find('#login-error-message').text()).toBe('Some error')
      })
    })
  })

  describe('when user is logged in', () => {
    it('should show only the logged in message', async () => {
      jwt_decode.mockReturnValue({ userId: '1234' })

      await createLoginForm()
      await wrapper.vm.$nextTick()

      await store.dispatch('auth/setToken', 'someToken')

      expect(wrapper.find('#form-login').exists()).toBeFalsy()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
