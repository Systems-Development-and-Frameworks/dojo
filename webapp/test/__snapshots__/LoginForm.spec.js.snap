// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm when no user is logged in should show the login form 1`] = `
"<div>
  <form id=\\"form-login\\">
    <table>
      <tbody>
        <tr>
          <td><label for=\\"input-email\\"> E-Mail </label></td>
          <td><input id=\\"input-email\\" type=\\"text\\" placeholder=\\"Enter your e-mail address..\\" required=\\"required\\"></td>
        </tr>
        <tr>
          <td><label for=\\"input-password\\"> Password </label></td>
          <td><input id=\\"input-password\\" type=\\"password\\" placeholder=\\"Enter your password..\\" required=\\"required\\"></td>
        </tr>
        <tr>
          <td colspan=\\"2\\"><input type=\\"submit\\" value=\\"Login\\" disabled=\\"disabled\\"></td>
        </tr>
        <!---->
        <!---->
      </tbody>
    </table>
  </form>
</div>"
`;

exports[`LoginForm when user is logged in should show only the logged in message 1`] = `
"<div><span>You're logged in!</span>
  <nuxtlink-stub to=\\"/\\">Return to home page</nuxtlink-stub>
  or
  <a href=\\"#\\">return to the previously browsed page</a></div>"
`;
